Add popup to DeepL bulk action of page tree view

JoeyStk opened the pull request:
### Short description
This PR adds the option to translate pages via MT through a bulk action inside the page tree view. Please note, this PR is not yet ready for review, since I still want to take the time to attempt refactoring it myself. This PR is supposed to fulfill the following use case: I, as CMS user, want to get an overview about my MT budget, the pages that will be able to be translated and those that can't be translated (due to too low HIX score), before translating them.


### Proposed changes

- Add release note
- Add a option for MT to the bulk action dropdown
- Add a popup to get a report about the attempted MT
  - Show from which to which language this translation is going to be
  - Show the pages that can be translated
  - Show the pages that can't be translated
  - Show the budget and calculate, if there is enough budget for it
- Request hix score and amount of words per page via AJAX  

### Background information
- After a chat with @osmers we decided to disable the option to switch the language inside the popup, since a. we don't want to encourage this behavior, and it would have increased the scope of this PR quite a bit
- After a chat with @hauf-toni we decided to keep most of the structure for the popup in comparison to the XLIFF popup, but add two significant changes: 1. remove the X to close the popup and instead adding a "Cancel" button, since it's more prominent and gives clearer instruction to the user of what to do and 2. widen the popup from 400px to 500px, because 400px were too narrow and added a lot of additional vertical scroll. 

### Side effects

- This could be effecting the MT via page, event or poi form
- There might be a bunch of merge conflicts after #2209 is merged
- Let's say you have 1 page that can be translated and one that can't be, you'll get two error messages for the page that can't be translated (one inside the popup via JS/TS and the second one after confirming via Dj/Py).
- I noticed, that currently I can re-translate a page via MT although the translation status is "Up to date" or "Machine translated and up to date". In order to be more user friendly and prevent unnecessary usage of MT budget, I would suggest adding a check if a machine translation is actually needed or not. First I think it's up for discussion if we really want this or not, and secondly I think that would also be out of scope for this PR and rather something for another issue  


### Resolved issues
<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #2102 


__________________________________________________
<!-- Keep this link for the potential reviewer -->
[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

charludo commented:
This is already looking fantastic! One issue I noticed (or maybe I'm just doing something wrong): The `Create missing translations automatically` bulk action is only shown to me on non-German page tree views, but selecting it then shows the error/warning about the too-low-for-translation HIX scores 😅

timoludwig reviewed (commented):
Wow, thanks a lot! :rocket:

The UI looks very cool! Before proceeding with the review, I think it would be good to resolve the conflicts. Since most of them are my fault, I could take care of that, or do you want to have a look yourself?

JoeyStk commented:
@timoludwig If you want to, feel free to resolve the conflicts. I'm a bit worried, that I might screw that up. If you don't find the time, I have a cautious attempt at it some time later this week

timoludwig commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,241 @@
+    document.getElementById("machine-translation-overlay-pages").innerHTML = "";
+    document.getElementById("machine-translation-overlay-pages-optional").innerHTML = "";
+    const selectedPages = getAllPages(getSelectedItems());
+    pageTitles.forEach((title, i) => {
+        if (selectedPages.includes(title)) {
+            const list = document.createElement("li");
+            list.textContent = title as string;
+            list.classList.add("list-disc");
+            list.classList.add("ml-4");
+            if (i <= 4) {
+                document.getElementById("machine-translation-overlay-pages").appendChild(list);
+            } else {
+                document.getElementById("machine-translation-overlay-pages-optional").appendChild(list);
+            }
+        }
+    });
```
I think building HTML elements dynamically in JS is a bit more complicated than needed - instead, I'd suggest to do the calculation which pages are allowed and which aren't in Python, as well as doing the rendering process in Django templates, and let the ajax view return the rendered HTML (see 
[`render_to_string()`](https://docs.djangoproject.com/en/3.2/topics/templates/#django.template.loader.render_to_string))...
```
return JsonResponse(
    {
        "all_pages": all_pages,
        "pages_to_be_translated": render_to_string(...),
        "pages_invalid": render_to_string(...),
    }
)
```

timoludwig replied:
On second thought, I think this is fine. I mean the JS implementation works, so probably it would be a waste of time to implement the same thing in Python. Just keep in mind this option for next time 😁

JoeyStk replied:
Ok, thank you. I'll write this down in my notes

timoludwig commented on integreat_cms/cms/views/pages/page_actions.py:
```diff
@@ -896,3 +896,54 @@ def refresh_date(
+    all_pages = {}
+    for language in request.region.active_languages:
+        default_language = request.region.default_language
+        if language == default_language:
```
Why this check? We also want to enable machine translations where the source language is not the default language.

JoeyStk replied:
I changed this a little bit now, can you please have a look if I went into the right direction?

timoludwig commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,241 @@
+};
+
+const setRequiredBudget = (sum: number) => {
+    document.getElementById("machine-translation-overlay-budget-usage-result").innerHTML = "";
```
```suggestion
    document.getElementById("machine-translation-overlay-budget-usage-result").textContent = "";
```
I'd suggest to use `textContent` instead of `innerHTML` whenever you're not really inserting raw HTML. This also applies to all other occurrences of `innerHTML`.

JoeyStk replied:
I tried to apply this as often as I thought it made sense. I keep innerHTML on two occasions on which it serves as a cleaner to get rid off the HTML child elements. I hope that decision was correct

timoludwig reviewed (changes requested):
Thanks a lot, really cool! :rocket: 

I know this wasn't ready for review yet, but I had a few preliminary comments nonetheless...

Also, I noticed two general things:
- The HIX value is also enforced when HIX is disabled in a region (or globally)... this does not make sense in my opinion, since we have different production systems where HIX might be disabled, but MTs should be available. Also, I think the budget warning could also be interesting for source languages where HIX is not enabled or events & pois which also don't have HIX enabled...
- I'm not completely sure why, but somehow the warning "None of the selected pages fullfills the required HIX score of 15." was shown for me, even when there exist pages with a sufficient HIX score... 


> Let's say you have 1 page that can be translated and one that can't be, you'll get two error messages for the page that can't be translated (one inside the popup via JS/TS and the second one after confirming via Dj/Py).

You could circumvent this by removing all invalid pages from the selected pages before submitting the form.

> I noticed, that currently I can re-translate a page via MT although the translation status is "Up to date" or "Machine translated and up to date". In order to be more user friendly and prevent unnecessary usage of MT budget, I would suggest adding a check if a machine translation is actually needed or not. First I think it's up for discussion if we really want this or not, and secondly I think that would also be out of scope for this PR and rather something for another issue

Yes good point, I suggest to open an issue for this.

JoeyStk commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,241 @@
+/* eslint no-magic-numbers: ["error", { "ignore": [0, 1, 3, 4, 5, 15] }] */
```
Oh, this is a  good point. Thank you! 
Would you prefer for me to get the variable from the settings (for example via a hidden HTML tag) or just use 15 as a constant for HIX?

timoludwig replied:
Yes, retrieving it from the settings and passing it to the template (e.g. as data-attribute) makes the most sense I think.

JoeyStk replied:
This should be applied now

JoeyStk commented:
Finally I took the time to work on this some more, and came up with the following changes. Technically speaking it'S not really refactoring, but I think it can be considered as cleaning up the code a bit:

- I split the setAmount() into two functions/methods, calculateAmount() for doing the logic and setAmount() for actually setting it to the HTML element
- As descibed above I tried to use textContent on as many occasions as I think made sense. There are still two where I use innerHTML since the HTML element contains children which should be cleaned by innerHTML.
- I created more constants to avoid the abundance of document.getElement I had before
- I tried to improve the names of variables and functions in order to make them more self-describing
- I added a comment to every function about what it is doing
- I removed hixThreshold as a new TS constant and instead access the variable we have in our setting
- I changed check if enough budget is remaining from larger than 0 to equal or larger than 0
- I reduced  the amount of magic numbers by setting constants and using them a bit more wisely. The three remaining magic numbers are either used as indexes or as increments. I thought it didn't make sense to also declare them as constants
- I repaired the budget variable (I screwed this up while rebasing my local setup)
- I repaired the "Show all pages" trigger, since it had a logical error
- I changed the Cancel button to be red, since this was a request of @osmers 
- I also added the number of selected pages to the overview, since also this was a request of @osmers 

There are still a few things I didn't do in this go
- I didn't change the names/ids of HTML elements, since I didn't want to screw anything up and I didn't have great alternatives in mind anyway. So please feel free to leave suggestions here
- @osmers asked if we/I also could add the option for the pages that don't have a sufficient hix score to open the editor of the according page by clicking on it's page title. Since I'm reaching the limits of my brain, I would suggest to open a new issue for this and add this feature once this PR has been merged

MizukiTemma reviewed (changes requested):
Thank you for PR 💪 

Sadly this problem still exists, even when the HIX is activated and the score is good enough...
>I'm not completely sure why, but somehow the warning "None of the selected pages fullfills the required HIX score of 15." was shown for me, even when there exist pages with a sufficient HIX score...

JoeyStk commented:
> Sadly this problem still exists, even when the HIX is activated and the score is good enough...
> 
> > I'm not completely sure why, but somehow the warning "None of the selected pages fullfills the required HIX score of 15." was shown for me, even when there exist pages with a sufficient HIX score...

Thank you for your reply. Hmm, strange. I can't reproduce this. I get the correct selection of pages, when I open the popup. Can you please give me some more details?
![grafik](https://user-images.githubusercontent.com/72705147/232822832-b9fc7480-ad84-487c-9b54-5590947eaf2b.png)

JoeyStk commented:
> Thanks a lot, really cool! rocket
> 
> I know this wasn't ready for review yet, but I had a few preliminary comments nonetheless...
> 
> Also, I noticed two general things:
> 
>     * The HIX value is also enforced when HIX is disabled in a region (or globally)... this does not make sense in my opinion, since we have different production systems where HIX might be disabled, but MTs should be available. Also, I think the budget warning could also be interesting for source languages where HIX is not enabled or events & pois which also don't have HIX enabled...
> 
>     * I'm not completely sure why, but somehow the warning "None of the selected pages fullfills the required HIX score of 15." was shown for me, even when there exist pages with a sufficient HIX score...
> 
> 
> > Let's say you have 1 page that can be translated and one that can't be, you'll get two error messages for the page that can't be translated (one inside the popup via JS/TS and the second one after confirming via Dj/Py).
> 
> You could circumvent this by removing all invalid pages from the selected pages before submitting the form.
> 
> > I noticed, that currently I can re-translate a page via MT although the translation status is "Up to date" or "Machine translated and up to date". In order to be more user friendly and prevent unnecessary usage of MT budget, I would suggest adding a check if a machine translation is actually needed or not. First I think it's up for discussion if we really want this or not, and secondly I think that would also be out of scope for this PR and rather something for another issue
> 
> Yes good point, I suggest to open an issue for this.

I noticed, that I missed that bit of feedback. I will have to take a closer look at it next week

JoeyStk commented:
> Yes good point, I suggest to open an issue for this.
[Done!] (https://github.com/digitalfabrik/integreat-cms/issues/2247)

JoeyStk commented:
Thanks for the feedback. I'll look into it! :smiley:

MizukiTemma commented:
@JoeyStk 
Sorry for direct push, the changes were too large to suggest as comment.
The overlay should show now translatable and untranslatable pages correctly. HIX check is now working with page IDs not titles.

The remaining ToDo are
1) Bipass HIX check if HIX is disabled in the region or globally.
2) Exclude pages that did not pass the HIX check from translation request.

JoeyStk commented:
@MizukiTemma 
Thank you again for your push. I think I found something for Todo#1 and I fixed an issue with the javascript toggle ("Show more" / "Show less"). But I think there is still the issue, that we have to get only the selected pages from the body via post. However I wasn't able to find a solution for that, because as far as I understand the issue the post body is empty as soon as the function post_hix_and_word_number_per_page is called. Do you maybe have an idea for that?

MizukiTemma commented:
@JoeyStk Thank you for updates. 
IDs of the selected pages are now passed to the ajax function as boddy data appended in the fetch and the process works only with the selected pages.
HIX bipass now checks not only the global deactivation but also the cases where HIX is globally activated but deactivated in the region too.

timoludwig commented on integreat_cms/cms/templates/pages/_page_machine_translation_overlay.html:
```diff
@@ -0,0 +1,101 @@
+                    <div class="my-2 bg-yellow-100 border-l-4 border-yellow-500 text-yellow-600 px-4 py-3 mb-5"
+                         role="alert">
+                        <i class="mr-2" icon-name="alert-triangle"></i>
+                        {% translate "There are currently" %} <span id="machine-translation-overlay-warning-number"></span> {% translate "pages that can't be translated. These pages can't be translated because the text understandibility is too low." %} {% translate "Please improve it to a value of at least 15 to translate these pages. You can identify the pages, that couldn't be translated in the ...
```
Also, I've explained quite often that it's bad practise to split up translatable sentences into multiple segments. There is no harm in using html markup inside translations, maybe shorten the id name a bit to make it better readable in the translation file.

timoludwig commented on integreat_cms/cms/templates/pages/_page_machine_translation_overlay.html:
```diff
@@ -0,0 +1,101 @@
+                    <div class="mt-2">
+                        <div id="machine-translation-overlay-pages-no-pages-warning"
+                             class="hidden bg-red-100 border-l-4 border-red-500 text-red-500 px-4 py-3 mb-5">
+                            {% translate "None of the selected pages fullfills the required HIX score of " %}  {{ hix_threshold }}.
```
Also here the variable should be included in the translation text since it's part of the same sentence.

timoludwig commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,267 @@
+
+const getSelectedItems = () => {
+    // get the HTML tr element of the selected pages
+    const checkboxes = document.querySelectorAll(".bulk-select-item");
```
```suggestion
    const checkboxes = document.querySelectorAll(".bulk-select-item:checked");
```
You can directly retrieve the selected checkboxes.

timoludwig commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,267 @@
+    const selectedPages: Array<number> = [];
+    checkboxes.forEach((checkbox) => {
+        if ((checkbox as HTMLInputElement).checked) {
+            selectedPages.push(parseInt(checkbox.closest("tr").getAttribute("data-drop-id"), 10));
```
Why go the complicated way via the `tr` and `data-drop-id` when you could just access the value of the checkbox directly?

timoludwig commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,267 @@
+
+const getSelectedIDs = () => {
+    // get the HTML tr element of the selected pages
+    const checkboxes = document.querySelectorAll(".bulk-select-item");
```
```suggestion
    const checkboxes = document.querySelectorAll(".bulk-select-item:checked");
```
You can filter the checked checkboxes directly. Also, why is this queryselector duplicated?

timoludwig commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,267 @@
+/* eslint no-magic-numbers: ["error", { "ignore": [0, 1, 3] }] */
+import { getCsrfToken } from "../utils/csrf-token";
+
+const limitOfPreview = 5;
```
If the limit is a variable, I would not hardcode the number in the variable name `listOfFirstFivePages`.

timoludwig commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,267 @@
+const toggleMachineTranslationForPagesButton = () => {
+    // Only activate button if budget is enough
+    const budget = document.getElementById("machine-translation-overlay-budget-remains-result")
+        .innerText as any as number;
```
Yes another occurrence of this pattern.

timoludwig commented on integreat_cms/static/src/js/pages/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,267 @@
+    const selectedPages: Array<Element> = [];
+    checkboxes.forEach((checkbox) => {
+        if ((checkbox as HTMLInputElement).checked) {
+            selectedPages.push(checkbox.closest("tr"));
```
I don't see a reason to do this complicated stuff with the additional array of `tr`s when you just need the number of selected pages anyway.

timoludwig commented on integreat_cms/cms/templates/pages/_page_machine_translation_overlay.html:
I don't see a reason to limit this to pages... why not make this generic for events and pois as well?

JoeyStk replied:
The service-team agrees with you on that. We'll extend the popup to events and pois as well

MizukiTemma replied:
In my opinion we can generalize the pop up in a separate issue/PR because it is beyond the original scope of issue and this PR is already kind of voluminous. We probably get some feedback and wishes from users for the page pop up and can apply them while generalizing.

timoludwig replied:
Fine for me 👍
Just move the files to a more general location, so they don't have to be moved later on.

JoeyStk commented:
Please note: This PR is not ready for re-review. I just pushed the recent changes, so @MizukiTemma can help me working on it

timoludwig commented on integreat_cms/cms/views/pages/page_actions.py:
```diff
@@ -896,3 +897,62 @@ def refresh_date(
+    not_translatable = {}
+
+    for page in selected_pages:
+        translation = page.get_translation(request.region.default_language.slug)
```
In my opinion, we should just skip the HIX value check in this case. I would be surprised if the HIX value of the German source text had a huge influence on the quality of machine translations from and to completely different languages.
Also, this would reflect the same constraints that we enforce in the forms (where you can enable machine translations for any language other than German independently from the HIX score).

JoeyStk commented:
Thank you for all your suggestions and explanations. The changes should be applied now. We double-checked each other to make sure we don't miss any comment or misinterpret it, so hopefully the implemented changes are what you had in mind. Please let us know if something is not as intended

MizukiTemma commented on integreat_cms/static/src/js/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,216 @@
+    expandableWarningListElement.textContent = "";
+    buildNotTranslatableList(filteredPages.not_translatable, expandableWarningListElement);
+
+    const expansionTrigger = document.getElementById("machine-translation-overlay-expansion-trigger");
+    if (previewListOfPages.children.length === 0) {
+        document.getElementById("machine-translation-overlay-pages-no-pages-warning").classList.remove("hidden");
+    } else if (previewListOfPages.children.length > limitOfPreview) {
+        expansionTrigger.classList.add("block");
+        expansionTrigger.classList.remove("hidden");
+    } else {
+        expansionTrigger.classList.add("hidden");
+        expansionTrigger.classList.remove("block");
+    }
+    toggleMachineTranslationForPagesButton();
+
+    if (expandableWarningListElement.children.length !== 0) {
+        warningListElement.classList.add("block");
+        warningListElement.classList.remove("hidden");
+    }
```
```suggestion
    const expansionTrigger = document.getElementById("machine-translation-overlay-expansion-trigger");
    const noPageWarning = document.getElementById("machine-translation-overlay-pages-no-pages-warning");

    if (previewListOfPages.children.length === 0) {
        noPageWarning.classList.remove("hidden");
    } else {
        noPageWarning.classList.add("hidden");
    }

    if (listOfOptionalPages.children.length === 0) {
        expansionTrigger.classList.add("hidden");
        expansionTrigger.classList.remove("block");
    } else {
        expansionTrigger.classList.remove("hidden");
        expansionTrigger.classList.add("block");
    }

    toggleMachineTranslationForPagesButton();

    if (expandableWarningListElement.children.length !== 0) {
        warningListElement.classList.add("block");
        warningListElement.classList.remove("hidden");
    } else {
        warningListElement.classList.remove("block");
        warningListElement.classList.add("hidden");
    }
```

The warning test "None of the selected pages fullfills the required HIX...." was not toggled correctlly.

MizukiTemma commented on integreat_cms/static/src/js/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,216 @@
+        list.classList.add("list-disc", "ml-4");
+        list.textContent = page.title;
+
+        if (previewListOfPages.children.length <= limitOfPreview) {
```
```suggestion
        if (previewListOfPages.children.length < limitOfPreview) {
```

MizukiTemma commented on integreat_cms/static/src/js/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,216 @@
+                filteredPages.then((filteredPages: Pages) => {
+                    prepareOverlay(filteredPages);
+                });
+
+                toggleOptionalText(
+                    document.getElementById("machine-translation-overlay-expansion-trigger"),
+                    document.getElementById("machine-translation-overlay-pages-optional")
+                );
+                toggleOptionalText(
+                    document.getElementById("machine-translation-overlay-warning-more"),
+                    document.getElementById("machine-translation-overlay-warning-optional")
+                );
+                setAmountOfSelectedPages(selectedIDs.length);
+            }
+        });
+        document.getElementById("btn-close-machine-translation-overlay").addEventListener("click", () => {
+            closeMachineTranslationOverlay(overlay);
+        });
+        // Close window by clicking on backdrop.
+        overlay.addEventListener("click", (e) => {
+            if (e.target === overlay) {
+                closeMachineTranslationOverlay(overlay);
+            }
+        });
+    }
```
```suggestion
                
                setAmountOfSelectedPages(selectedIDs.length);
            }
        });
        document.getElementById("btn-close-machine-translation-overlay").addEventListener("click", () => {
            closeMachineTranslationOverlay(overlay);
        });
        // Close window by clicking on backdrop.
        overlay.addEventListener("click", (e) => {
            if (e.target === overlay) {
                closeMachineTranslationOverlay(overlay);
            }
        });

        toggleOptionalText(
            document.getElementById("machine-translation-overlay-warning-more"),
            document.getElementById("machine-translation-overlay-warning-optional")
        );
        toggleOptionalText(
            document.getElementById("machine-translation-overlay-expansion-trigger"),
            document.getElementById("machine-translation-overlay-pages-optional")
        );
    }
```

Event listner on "Show pages" were piled up and click toggle didn't work sometimes.

MizukiTemma reviewed (changes requested):
Thank you for rebase and update 💪 

I've noticed some incorrect behaviour on toggling of the warning message and click event listner of showing more/less pages.

After those changes the PR is ready for meage in my opinion 🙂

JoeyStk commented:
Thank you for your suggestions. They are applied now :)

MizukiTemma reviewed (approved):
🎉

timoludwig commented on integreat_cms/cms/templates/_machine_translation_overlay.html:
```diff
@@ -0,0 +1,117 @@
+                         role="alert">
+                        <i class="mr-2" icon-name="alert-triangle"></i>
+                        {% blocktranslate trimmed %}
+                            There are currently <span id="not-translateable"></span> page(s) that can't be translated.
```
```suggestion
                            There are currently <span id="not-translatable"></span> page(s) that can't be translated.
```
Small typo

timoludwig commented on integreat_cms/locale/de/LC_MESSAGES/django.po:
```diff
@@ -3951,6 +3951,125 @@ msgstr "Standard-Sprache wird dupliziert"
+
+#: cms/templates/_machine_translation_overlay.html
+msgid ""
+"There are currently <span id=\"not-translateable\"></span> page(s) that "
+"can't be translated."
+msgstr ""
+"Es gibt aktuell <span id=\"not-translateable\"></span> Seiten, die nicht "
```
```suggestion
"There are currently <span id=\"not-translatable\"></span> page(s) that "
"can't be translated."
msgstr ""
"Es gibt aktuell <span id=\"not-translatable\"></span> Seiten, die nicht "
```

timoludwig commented on integreat_cms/static/src/js/machine-translation-overlay.ts:
```diff
@@ -0,0 +1,225 @@
+        numberOfNotTranslatablePages += 1;
+    });
+
+    document.getElementById("not-translateable").textContent = numberOfNotTranslatablePages.toString();
```
```suggestion
    document.getElementById("not-translatable").textContent = numberOfNotTranslatablePages.toString();
```

timoludwig reviewed (changes requested):
Thanks a lot, yes this looks a lot better now! :+1:

JoeyStk commented on integreat_cms/cms/templates/_machine_translation_overlay.html:
```diff
@@ -0,0 +1,117 @@
+                        {% blocktranslate trimmed %}
+                            There are currently <span id="not-translateable"></span> page(s) that can't be translated.
+                        {% endblocktranslate %}
+                        {% if source_language.slug == "de" %}
```
This should be applied now

JoeyStk commented:
Thank you a lot for your feedback. The changes should be applied now, the PR should be rebased to only one commit and no merge conflict should be open anymore. With that I think this PR is ready for re-review :)

timoludwig reviewed (approved):
Thanks a lot! :rocket:
